src/tallies/tally_filter.cpp

```diff
@@ -0,0 +1,561 @@
+    sph_filt->order_ = order;
+    sph_filt->n_bins_ = (order + 1) * (order + 1);
+    filter_update_n_bins(index);
+    return 0;
+  }
+
+  int
+  openmc_sphharm_filter_set_cosine(int32_t index, const char cosine[])
+  {
+    int err = verify_filter(index);
+    if (err) return err;
+
+    auto filt = filter_from_f(index);
+    if (filt->type() != "sphericalharmonics") {
+      set_errmsg("Not a spherical harmonics filter.");
+      return OPENMC_E_INVALID_TYPE;
+    }
+
+    auto sph_filt = static_cast<SphericalHarmonicsFilter*>(filt);
+    if (strcmp(cosine, "scatter") == 0) {
+      sph_filt->cosine_ = SphericalHarmonicsCosine::scatter;
+    } else if (strcmp(cosine, "particle") == 0) {
+      sph_filt->cosine_ = SphericalHarmonicsCosine::particle;
+    } else {
+      set_errmsg("Invalid spherical harmonics cosine.");
+      return OPENMC_E_INVALID_ARGUMENT;
+    }
+    return 0;
+  }
+
+  int
+  openmc_zernike_filter_get_order(int32_t index, int* order)
+  {
+    int err = verify_filter(index);
+    if (err) return err;
+
+    auto filt = filter_from_f(index);
+    if (filt->type() != "zernike" && filt->type() != "zernikeradial") {
+      set_errmsg("Not a Zernike filter.");
+      return OPENMC_E_INVALID_TYPE;
+    }
+
+    auto z_filt = static_cast<ZernikeFilter*>(filt);
+    *order = z_filt->order_;
+    return 0;
+  }
+
+  int
+  openmc_zernike_filter_get_params(int32_t index, double* x, double* y,
+                                   double* r)
+  {
+    int err = verify_filter(index);
+    if (err) return err;
+
+    auto filt = filter_from_f(index);
+    if (filt->type() != "zernike" && filt->type() != "zernikeradial") {
+      set_errmsg("Not a Zernike filter.");
+      return OPENMC_E_INVALID_TYPE;
+    }
+
+    auto z_filt = static_cast<ZernikeFilter*>(filt);
+    *x = z_filt->x_;
+    *y = z_filt->y_;
+    *r = z_filt->r_;
+    return 0;
+  }
+
+  int
+  openmc_zernike_filter_set_order(int32_t index, int order)
+  {
+    int err = verify_filter(index);
+    if (err) return err;
+
+    auto filt = filter_from_f(index);
+    if (filt->type() != "zernike" && filt->type() != "zernikeradial") {
+      set_errmsg("Not a Zernike filter.");
+      return OPENMC_E_INVALID_TYPE;
+    }
+
+    auto z_filt = static_cast<ZernikeFilter*>(filt);
+    z_filt->order_ = order;
+    z_filt->calc_n_bins();
+    filter_update_n_bins(index);
+    return 0;
+  }
+
+  int
+  openmc_zernike_filter_set_params(int32_t index, const double* x,
+                                   const double* y, const double* r)
+  {
+    int err = verify_filter(index);
+    if (err) return err;
+
+    auto filt = filter_from_f(index);
+    if (filt->type() != "zernike" && filt->type() != "zernikeradial") {
+      set_errmsg("Not a Zernike filter.");
+      return OPENMC_E_INVALID_TYPE;
+    }
+
+    auto z_filt = static_cast<ZernikeFilter*>(filt);
+    if (x) z_filt->x_ = *x;
+    if (y) z_filt->y_ = *y;
+    if (r) z_filt->r_ = *r;
+    return 0;
+  }
+}
```

Comment:
If we do split out tally filter implementations into separate files, the C API functions for specific filters can go into the respective .cpp files.